Bot.Core/TownResolver.cs

```diff
@@ -0,0 +1,142 @@
+﻿using Bot.Api;
+using Bot.Api.Database;
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Bot.Core
+{
+    public class TownResolver : ITownResolver
+    {
+        private readonly IBotClient m_client;
+        private readonly ITownDatabase m_townDb;
+
+        public TownResolver(IServiceProvider serviceProvider)
+        {
+            serviceProvider.Inject(out m_client);
+            serviceProvider.Inject(out m_townDb);
+        }
+
+        public async Task<ITown?> ResolveTownAsync(ITownRecord rec)
+        {
+            var guild = await m_client.GetGuildAsync(rec.GuildId);
+            if (guild != null)
+            {
+                var controlChannelResult = GetChannel(guild, rec.ControlChannelId, rec.ControlChannel, false);
+                var dayCategoryResult = GetChannelCategory(guild, rec.DayCategoryId, rec.DayCategory);
+                var nightCategoryResult = GetChannelCategory(guild, rec.NightCategoryId, rec.NightCategory);
+                var chatChannelResult = GetChannel(guild, rec.ChatChannelId, rec.ChatChannel, false);
+                var townSquareResult = GetChannel(guild, rec.TownSquareId, rec.TownSquare, true);
+
+                var town = new Town(rec)
+                {
+                    Guild = guild,
+                    ControlChannel = controlChannelResult.Channel,
+                    DayCategory = dayCategoryResult.Channel,
+                    NightCategory = nightCategoryResult.Channel,
+                    ChatChannel = chatChannelResult.Channel,
+                    TownSquare = townSquareResult.Channel,
+                    StorytellerRole = GetRoleForGuild(guild, rec.StorytellerRoleId),
+                    VillagerRole = GetRoleForGuild(guild, rec.VillagerRoleId),
+                };
+
+                if (AnyUpdatesRequired(
+                    controlChannelResult,
+                    dayCategoryResult,
+                    nightCategoryResult, 
+                    chatChannelResult,
+                    townSquareResult))
+                    await m_townDb.UpdateTownAsync(town);
+
+                return town;
+            }
+            return null;
+        }
+
+        private static bool AnyUpdatesRequired(params GetChannelResultBase[] results) => results.Any(r => r.UpdateRequired != ChannelUpdateRequired.None);
+
+        private static GetChannelResult GetChannel(IGuild guild, ulong channelId, string? channelName, bool expectedIsVoice)
+        {
+            ChannelUpdateRequired update = ChannelUpdateRequired.None;
+
+            var channel = guild.GetChannel(channelId);
+            if (channel == null)
+                channel = guild.Channels.FirstOrDefault(c => c.Name == channelName);
```

Comment:
You think maybe this should check the Day Category rather than the guild as a whole?